File: clients/typescript/diffdeltaClient.ts
/**
 * DiffDelta Reference Client — TypeScript
 *
 * Zero external dependencies.  Uses the standard `fetch` API (Node 18+ / browsers).
 * Implements the DiffDelta Feed Spec v1.1 polling protocol.
 *
 * **Key behavior:** Every call to `poll()` or `fetchLatest()` goes through
 * head.json first, sends `If-None-Match` with the cached ETag, and only
 * fetches the full payload when content has actually changed.
 * This is not optional — it's how the protocol scales.
 *
 * @example
 * ```ts
 * const client = new DiffDeltaClient("https://diffdelta.io");
 *
 * // Recommended: use poll() — head-first by default
 * const feed = await client.poll("aws_whats_new");
 * if (feed) {
 *   for (const item of feed.buckets.new) console.log(item.headline);
 * }
 * ```
 */

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface HeadPointer {
  cursor: string | null;
  prev_cursor: string | null;
  changed: boolean;
  generated_at: string;
  ttl_sec: number;
  latest_url: string;
}

export interface SourceStatus {
  name?: string;
  changed: boolean;
  cursor: string | null;
  prev_cursor: string | null;
  ttl_sec: number;
  status: "ok" | "error" | "disabled" | "unknown";
  delta_counts?: { new: number; updated: number; removed: number };
  stale?: boolean;
  stale_since?: string;
  stale_age_sec?: number;
  last_ok_at?: string;
  error?: { code: string; http_status?: number };
  disabled_reason?: string;
  latest_url?: string;
  head_url?: string;
}

export interface DeltaItem {
  source: string;
  id: string;
  url: string;
  headline: string;
  published_at: string;
  updated_at: string;
  /** Optional — omitted when score==0 and no reasons.  Treat missing as {score:0}. */
  risk?: { score: number; reasons?: string[] };
  provenance: {
    fetched_at: string;
    evidence_urls: string[];
    content_hash: string;
  };
  content: { lang: string; excerpt_text?: string };
  item_hash?: string;
  [key: string]: unknown;
}

export interface Feed {
  schema_version: string;
  generated_at: string;
  cursor: string | null;
  prev_cursor: string | null;
  changed: boolean;
  ttl_sec: number;
  sources_included: string[];
  sources: Record<string, SourceStatus>;
  buckets: {
    new: DeltaItem[];
    updated: DeltaItem[];
    removed: DeltaItem[];
    flagged: DeltaItem[];
    [key: string]: DeltaItem[];
  };
  batch_narrative: string;
  cursor_basis?: string;
  _discovery?: Record<string, string>;
  [key: string]: unknown;
}

export interface FetchHeadResult {
  /** True if there is new semantic content since the last poll. */
  changed: boolean;
  /** Head pointer JSON, or null on 304. */
  head: HeadPointer | null;
}

// ---------------------------------------------------------------------------
// In-memory cursor / ETag cache
// ---------------------------------------------------------------------------

class CursorCache {
  private data: Map<string, string> = new Map();

  get(key: string): string | undefined {
    return this.data.get(key);
  }

  set(key: string, value: string): void {
    this.data.set(key, value);
  }
}

// ---------------------------------------------------------------------------
// Client
// ---------------------------------------------------------------------------

const USER_AGENT = "diffdelta-ts-client/0.2.0";

export class DiffDeltaClient {
  private baseUrl: string;
  private timeout: number;
  private cache: CursorCache;

  /**
   * @param baseUrl  Origin of the DiffDelta server.
   * @param timeout  HTTP timeout in milliseconds (default 15 000).
   */
  constructor(baseUrl = "https://diffdelta.io", timeout = 15_000) {
    this.baseUrl = baseUrl.replace(/\/+$/, "");
    this.timeout = timeout;
    this.cache = new CursorCache();
  }

  // -- HTTP helpers --------------------------------------------------------

  private async httpGet<T>(
    url: string,
    etag?: string
  ): Promise<{ status: number; body: T | null; etag: string | null }> {
    const fullUrl = url.startsWith("http")
      ? url
      : `${this.baseUrl}${url.startsWith("/") ? "" : "/"}${url}`;

    const headers: Record<string, string> = {
      "User-Agent": USER_AGENT,
      Accept: "application/json",
    };
    if (etag) {
      headers["If-None-Match"] = `"${etag}"`;
    }

    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.timeout);

    try {
      const resp = await fetch(fullUrl, {
        headers,
        signal: controller.signal,
      });

      const respEtag = (resp.headers.get("etag") ?? "").replace(/"/g, "");

      if (resp.status === 304) {
        return { status: 304, body: null, etag: respEtag };
      }

      if (!resp.ok) {
        throw new Error(`HTTP ${resp.status} from ${fullUrl}`);
      }

      const body = (await resp.json()) as T;
      return { status: resp.status, body, etag: respEtag };
    } finally {
      clearTimeout(timer);
    }
  }

  // -- public API ----------------------------------------------------------

  /**
   * Poll the head pointer for a source.
   *
   * Sends `If-None-Match` with the locally cached ETag.
   * Returns `{ changed: false, head: null }` on 304.
   */
  async fetchHead(sourceId: string): Promise<FetchHeadResult> {
    const cacheKey = `etag:${sourceId}`;
    const storedEtag = this.cache.get(cacheKey);

    const url = `/diff/${sourceId}/head.json`;
    const { status, body, etag } = await this.httpGet<HeadPointer>(
      url,
      storedEtag
    );

    if (status === 304) {
      return { changed: false, head: null };
    }

    if (etag) {
      this.cache.set(cacheKey, etag);
    }

    const changed = body?.changed ?? true;
    return { changed, head: body };
  }

  /**
   * Head-first poll for a source.  Returns the feed or null.
   *
   * **This is the recommended entry point.**  It:
   * 1. Checks head.json (with If-None-Match).
   * 2. Returns `null` immediately on 304 or `changed: false`.
   * 3. Fetches latest.json only when content has actually changed.
   */
  async poll(sourceId: string): Promise<Feed | null> {
    const { changed, head } = await this.fetchHead(sourceId);
    if (!changed || !head) return null;
    if (!head.changed) return null;

    const latestUrl = head.latest_url ?? `/diff/${sourceId}/latest.json`;
    return this.fetchDirect<Feed>(latestUrl);
  }

  /**
   * Head-first poll for the global aggregated feed.
   * Returns the full feed, or null if nothing changed.
   */
  async pollGlobal(): Promise<Feed | null> {
    const cacheKey = "etag:_global";
    const storedEtag = this.cache.get(cacheKey);

    const { status, body, etag } = await this.httpGet<HeadPointer>(
      "/diff/head.json",
      storedEtag
    );

    if (status === 304) return null;

    if (etag) {
      this.cache.set(cacheKey, etag);
    }

    if (!(body?.changed ?? true)) return null;

    return this.fetchDirect<Feed>("/diff/latest.json");
  }

  /**
   * Fetch latest feed for a source, head-first.
   *
   * Equivalent to `poll()` — exists for backward compatibility.
   */
  async fetchLatest(sourceId: string): Promise<Feed | null> {
    return this.poll(sourceId);
  }

  /** Fetch the global aggregated feed, head-first. */
  async fetchGlobal(): Promise<Feed | null> {
    return this.pollGlobal();
  }

  /**
   * Fetch a URL directly (no head check).
   *
   * Use ONLY when you already know content has changed
   * (e.g. after walking the archive).
   */
  async fetchDirect<T>(url: string): Promise<T> {
    const { body } = await this.httpGet<T>(url);
    if (!body) throw new Error(`Empty response from ${url}`);
    return body;
  }

  /**
   * Walk the archive chain for historical snapshots.
   *
   * Fetches `/archive/{sourceId}/index.json` and retrieves up to
   * `limit` snapshots (newest first).  For onboarding / catchup only.
   */
  async walkBack(sourceId: string, limit = 10): Promise<Feed[]> {
    const snapshots: Feed[] = [];

    let index: { snapshots?: Array<{ url?: string }> };
    try {
      index = await this.fetchDirect(`/archive/${sourceId}/index.json`);
    } catch {
      return snapshots;
    }

    const entries = index.snapshots ?? [];
    const recent = entries.slice(-limit).reverse();

    for (const entry of recent) {
      if (!entry.url) continue;
      try {
        const snap = await this.fetchDirect<Feed>(entry.url);
        snapshots.push(snap);
      } catch {
        break;
      }
    }

    return snapshots;
  }
}

// ---------------------------------------------------------------------------
// CLI demo (Node.js)
// ---------------------------------------------------------------------------

async function main(): Promise<void> {
  const baseUrl = process.env.DIFFDELTA_BASE_URL ?? "https://diffdelta.io";
  const sourceId = process.argv[2] ?? "aws_whats_new";

  const client = new DiffDeltaClient(baseUrl);

  console.log(`Polling ${baseUrl} for source '${sourceId}' …`);
  const feed = await client.poll(sourceId);

  if (!feed) {
    console.log("Nothing new (304 or changed:false). Done.");
    return;
  }

  console.log(`Changed! Cursor: ${(feed.cursor ?? "null").slice(0, 24)}…`);

  for (const bucket of ["new", "updated", "removed", "flagged"] as const) {
    const items = feed.buckets[bucket] ?? [];
    if (items.length > 0) {
      console.log(`\n  [${bucket}] ${items.length} item(s):`);
      for (const item of items.slice(0, 5)) {
        const flag = (item.risk?.score ?? 0) >= 0.4 ? " ⚠" : "";
        console.log(`    • ${item.headline ?? "(no headline)"}${flag}`);
      }
      if (items.length > 5) {
        console.log(`    … and ${items.length - 5} more`);
      }
    }
  }

  console.log("\nDone.");
}

// Run if executed directly (Node / tsx / ts-node)
main().catch(console.error);

File: clients/python/diffdelta_client.py
#!/usr/bin/env python3
"""
DiffDelta Reference Client — Python

Zero external dependencies. Uses only the Python standard library.
Implements the DiffDelta Feed Spec v1.1 polling protocol.

**Key behavior:** Every call to ``poll()`` or ``fetch_latest()`` goes
through head.json first, sends ``If-None-Match`` with the cached ETag,
and only fetches the full payload when content has actually changed.
This is not optional — it's how the protocol scales.  At 100,000 bots,
skipping head.json turns a 200-byte 304 into a 50 KB full fetch.

Usage:
    from diffdelta_client import DiffDeltaClient

    client = DiffDeltaClient("https://diffdelta.io")

    # Recommended: use poll() — it does head-first automatically
    result = client.poll("aws_whats_new")
    if result is not None:
        for item in result["buckets"]["new"]:
            print(item["headline"])
"""

import hashlib
import json
import os
import sys
from pathlib import Path
from typing import Any, Dict, List, Optional, Tuple
from urllib.request import Request, urlopen
from urllib.error import HTTPError

__version__ = "0.2.0"

# ---------------------------------------------------------------------------
# Cache
# ---------------------------------------------------------------------------

_DEFAULT_CACHE_DIR = Path.home() / ".cache" / "diffdelta"


class CursorCache:
    """Persists per-source ETags and cursors to a local JSON file."""

    def __init__(self, cache_dir: Optional[Path] = None) -> None:
        self._dir = cache_dir or _DEFAULT_CACHE_DIR
        self._path = self._dir / "cursor_cache.json"
        self._data: Dict[str, str] = {}
        self._load()

    # -- persistence --------------------------------------------------------

    def _load(self) -> None:
        if self._path.exists():
            try:
                self._data = json.loads(self._path.read_text("utf-8"))
            except (json.JSONDecodeError, OSError):
                self._data = {}

    def _save(self) -> None:
        self._dir.mkdir(parents=True, exist_ok=True)
        tmp = self._path.with_suffix(".tmp")
        tmp.write_text(json.dumps(self._data, indent=2), "utf-8")
        tmp.replace(self._path)

    # -- public API ---------------------------------------------------------

    def get(self, key: str) -> Optional[str]:
        return self._data.get(key)

    def set(self, key: str, value: str) -> None:
        self._data[key] = value
        self._save()


# ---------------------------------------------------------------------------
# Client
# ---------------------------------------------------------------------------

_USER_AGENT = f"diffdelta-python-client/{__version__}"


class DiffDeltaClient:
    """Minimal DiffDelta polling client with ETag/304 support.

    The client enforces the two-step polling pattern by default:

    1. Fetch ``head.json`` with ``If-None-Match``.
    2. If 304 → done.  If ``changed: false`` → done.
    3. Only then fetch ``latest.json``.

    This behavior is built into ``poll()`` and ``fetch_latest()``.
    Bot operators don't need to think about it.

    Parameters
    ----------
    base_url : str
        Origin of the DiffDelta server (e.g. ``https://diffdelta.io``).
    cache_dir : Path | None
        Directory for persisting cursors/ETags.  Defaults to
        ``~/.cache/diffdelta``.
    timeout : int
        HTTP timeout in seconds.
    """

    def __init__(
        self,
        base_url: str = "https://diffdelta.io",
        cache_dir: Optional[Path] = None,
        timeout: int = 15,
    ) -> None:
        self.base_url = base_url.rstrip("/")
        self.timeout = timeout
        self._cache = CursorCache(cache_dir)

    # -- HTTP helpers -------------------------------------------------------

    def _get(
        self, url: str, etag: Optional[str] = None
    ) -> Tuple[int, Optional[Dict[str, Any]], Optional[str]]:
        """HTTP GET with optional If-None-Match.

        Returns (status, body_dict_or_None, etag_from_response).
        """
        if not url.startswith("http"):
            url = self.base_url + ("" if url.startswith("/") else "/") + url

        headers: Dict[str, str] = {
            "User-Agent": _USER_AGENT,
            "Accept": "application/json",
        }
        if etag:
            headers["If-None-Match"] = f'"{etag}"'

        req = Request(url, headers=headers)
        try:
            resp = urlopen(req, timeout=self.timeout)  # noqa: S310
            body = json.loads(resp.read().decode("utf-8"))
            resp_etag = (resp.headers.get("ETag") or "").strip('" ')
            return resp.status, body, resp_etag
        except HTTPError as exc:
            if exc.code == 304:
                resp_etag = (exc.headers.get("ETag") or "").strip('" ')
                return 304, None, resp_etag
            raise

    # -- public API ---------------------------------------------------------

    def fetch_head(
        self, source_id: str
    ) -> Tuple[bool, Optional[Dict[str, Any]]]:
        """Poll the head pointer for *source_id*.

        Uses a locally cached ETag to send ``If-None-Match``.
        Returns ``(changed, head_json_or_None)``.

        * ``changed=False, head=None`` → 304, nothing new.
        * ``changed=True,  head={…}``  → new data available.
        * ``changed=False, head={…}``  → fetched but no semantic change.
        """
        cache_key = f"etag:{source_id}"
        stored_etag = self._cache.get(cache_key)

        url = f"/diff/source/{source_id}/head.json"
        status, body, resp_etag = self._get(url, etag=stored_etag)

        if status == 304:
            return False, None

        # Persist ETag for next poll
        if resp_etag:
            self._cache.set(cache_key, resp_etag)

        changed = (body or {}).get("changed", True)
        return changed, body

    def poll(
        self, source_id: str
    ) -> Optional[Dict[str, Any]]:
        """Head-first poll for a source.  Returns the feed or None.

        This is the recommended entry point.  It:
        1. Checks head.json (with If-None-Match).
        2. Returns ``None`` immediately on 304 or ``changed: false``.
        3. Fetches latest.json only when content has actually changed.

        Returns the full feed dict, or ``None`` if nothing changed.
        """
        changed, head = self.fetch_head(source_id)
        if not changed or head is None:
            return None
        if not head.get("changed", True):
            return None

        latest_url = head.get("latest_url", f"/diff/source/{source_id}/latest.json")
        return self._fetch_json(latest_url)

    def poll_global(self) -> Optional[Dict[str, Any]]:
        """Head-first poll for the global aggregated feed.

        Returns the full feed dict, or ``None`` if nothing changed.
        """
        cache_key = "etag:_global"
        stored_etag = self._cache.get(cache_key)

        status, body, resp_etag = self._get("/diff/head.json", etag=stored_etag)

        if status == 304:
            return None

        if resp_etag:
            self._cache.set(cache_key, resp_etag)

        changed = (body or {}).get("changed", True)
        if not changed:
            return None

        return self._fetch_json("/diff/latest.json")

    def fetch_latest(self, source_id: str) -> Optional[Dict[str, Any]]:
        """Fetch latest feed for *source_id*, head-first.

        Equivalent to ``poll(source_id)`` — exists for backward compat.
        """
        return self.poll(source_id)

    def fetch_latest_direct(self, url: str) -> Dict[str, Any]:
        """Fetch a feed URL directly (no head check).

        Use this ONLY when you already know the content has changed
        (e.g. after walking the archive).
        """
        return self._fetch_json(url)

    def fetch_global(self) -> Optional[Dict[str, Any]]:
        """Fetch the global aggregated feed, head-first."""
        return self.poll_global()

    def fetch_sources(
        self, tags: Optional[List[str]] = None
    ) -> List[Dict[str, Any]]:
        """Fetch the source catalog and optionally filter by tags.

        Returns a list of source entries from ``/diff/sources.json``.
        If *tags* is provided, only sources matching **any** of the
        given tags are returned (OR logic).

        This is a setup-time call, not a polling call.  Cache the result
        and use the returned ``source_id`` values with ``poll()``.

        Example::

            # Get only security sources
            security = client.fetch_sources(tags=["security"])
            for src in security:
                feed = client.poll(src["source_id"])
        """
        data = self._fetch_json("/diff/sources.json")
        sources = data.get("sources", [])
        if tags:
            tag_set = set(tags)
            sources = [
                s for s in sources
                if tag_set.intersection(s.get("tags", []))
            ]
        return sources

    def walk_back(
        self,
        source_id: str,
        limit: int = 10,
    ) -> List[Dict[str, Any]]:
        """Walk the archive chain for historical snapshots.

        Fetches ``/archive/{source_id}/index.json`` and retrieves up to
        *limit* snapshots (newest first).  This is for onboarding / catchup,
        not for steady-state polling.
        """
        snapshots: List[Dict[str, Any]] = []

        try:
            index = self._fetch_json(f"/archive/{source_id}/index.json")
        except Exception:
            return snapshots

        entries = index.get("snapshots", [])
        # Newest first
        for entry in entries[-limit:][::-1]:
            url = entry.get("url")
            if not url:
                continue
            try:
                snap = self._fetch_json(url)
                snapshots.append(snap)
            except Exception:
                break

        return snapshots

    # -- internal -----------------------------------------------------------

    def _fetch_json(self, url: str) -> Dict[str, Any]:
        """Fetch a JSON URL directly (no ETag logic)."""
        status, body, _ = self._get(url)
        if body is None:
            raise RuntimeError(f"Empty response from {url}")
        return body


# ---------------------------------------------------------------------------
# CLI demo
# ---------------------------------------------------------------------------

def main() -> None:
    """Demonstrate the head-first polling loop."""
    base = os.environ.get("DIFFDELTA_BASE_URL", "https://diffdelta.io")
    source = sys.argv[1] if len(sys.argv) > 1 else "aws_whats_new"

    client = DiffDeltaClient(base)

    print(f"Polling {base} for source '{source}' …")
    feed = client.poll(source)

    if feed is None:
        print("Nothing new (304 or changed:false). Done.")
        return

    print(f"Changed! Cursor: {(feed.get('cursor') or 'null')[:24]}…")

    for bucket in ("new", "updated", "removed", "flagged"):
        items = feed.get("buckets", {}).get(bucket, [])
        if items:
            print(f"\n  [{bucket}] {len(items)} item(s):")
            for item in items[:5]:
                risk = (item.get("risk") or {}).get("score", 0)
                flag = " ⚠" if risk >= 0.4 else ""
                print(f"    • {item.get('headline', '(no headline)')}{flag}")
            if len(items) > 5:
                print(f"    … and {len(items) - 5} more")

    print("\nDone.")


if __name__ == "__main__":
    main()
